lib/resourcebuilder/apps.go

```diff
@@ -113,102 +111,68 @@ func (b *deploymentBuilder) Do(ctx context.Context) error {
-			lastErr = &payload.UpdateError{
-				Nested:  fmt.Errorf("deployment %s is not available; updated replicas=%d of %d, available replicas=%d of %d", iden, d.Status.UpdatedReplicas, d.Status.Replicas, d.Status.AvailableReplicas, d.Status.Replicas),
-				Reason:  "WorkloadNotAvailable",
-				Message: fmt.Sprintf("deployment %s is not available %s: %s", iden, availableCondition.Reason, availableCondition.Message),
-				Name:    iden,
-			}
-			return false, nil
-		}
-
-		if progressingCondition != nil && progressingCondition.Status == corev1.ConditionFalse && progressingCondition.Reason == "ProgressDeadlineExceeded" {
-			lastErr = &payload.UpdateError{
-				Nested:  fmt.Errorf("deployment %s is not progressing; updated replicas=%d of %d, available replicas=%d of %d", iden, d.Status.UpdatedReplicas, d.Status.Replicas, d.Status.AvailableReplicas, d.Status.Replicas),
-				Reason:  "WorkloadNotAvailable",
-				Message: fmt.Sprintf("deployment %s is not progressing %s: %s", iden, progressingCondition.Reason, progressingCondition.Message),
-				Name:    iden,
-			}
-			return false, nil
+	if replicaFailureCondition != nil && replicaFailureCondition.Status == corev1.ConditionTrue {
+		return &payload.UpdateError{
+			Nested:  fmt.Errorf("deployment %s has some pods failing; unavailable replicas=%d", iden, d.Status.UnavailableReplicas),
+			Reason:  "WorkloadNotProgressing",
+			Message: fmt.Sprintf("deployment %s has a replica failure %s: %s", iden, replicaFailureCondition.Reason, replicaFailureCondition.Message),
+			Name:    iden,
 		}
+	}
 
-		if progressingCondition != nil && progressingCondition.Status == corev1.ConditionTrue {
-			klog.V(4).Infof("deployment %s is progressing", iden)
-			return false, nil
+	if availableCondition != nil && availableCondition.Status == corev1.ConditionFalse {
+		return &payload.UpdateError{
+			Nested:  fmt.Errorf("deployment %s is not available; updated replicas=%d of %d, available replicas=%d of %d", iden, d.Status.UpdatedReplicas, d.Status.Replicas, d.Status.AvailableReplicas, d.Status.Replicas),
+			Reason:  "WorkloadNotAvailable",
+			Message: fmt.Sprintf("deployment %s is not available %s: %s", iden, availableCondition.Reason, availableCondition.Message),
+			Name:    iden,
 		}
+	}
 
-		klog.Errorf("deployment %s is in unknown state", iden)
-		return false, nil
-	}, ctx.Done())
-	if err != nil {
-		if err == wait.ErrWaitTimeout && lastErr != nil {
-			return lastErr
+	if progressingCondition != nil && progressingCondition.Status == corev1.ConditionFalse {
+		return &payload.UpdateError{
+			Nested:  fmt.Errorf("deployment %s is not progressing; updated replicas=%d of %d, available replicas=%d of %d", iden, d.Status.UpdatedReplicas, d.Status.Replicas, d.Status.AvailableReplicas, d.Status.Replicas),
+			Reason:  "WorkloadNotAvailable",
+			Message: fmt.Sprintf("deployment %s is not progressing %s: %s", iden, progressingCondition.Reason, progressingCondition.Message),
+			Name:    iden,
 		}
-		return err
 	}
+
```

Comment:
So with this change we don't care why we're not progressing, only that we're not, at this point in the process. I understand that to be in line with the overall goal of this change but would it not still be useful to post a log if we were in an `unknown state`? Would it be at all useful in debugging a deployment failure that bubbles up?

Reply:
> So with this change we don't care why we're not progressing, only that we're not...

Did we ever?

```console
$ git show --oneline -w c2af13b -- lib/resourcebuilder/apps.go
...
-func waitForDeploymentCompletion(ctx context.Context, client appsclientv1.DeploymentsGetter, deployment *appsv1.Deployment) error {
+
+func checkDeploymentHealth(ctx context.Context, client appsclientv1.DeploymentsGetter, deployment *appsv1.Deployment) error {
        iden := fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name)
-       var lastErr error
-       err := wait.PollImmediateUntil(defaultObjectPollInterval, func() (bool, error) {
...
-               if d.Generation <= d.Status.ObservedGeneration && d.Status.UpdatedReplicas == d.Status.Replicas && d.Status.UnavailableReplicas == 0 {
-                       return true, nil
...
```

The main changes I'm making here are removing the `PollImmediateUntil` (because we are no longer waiting, per the single-shot motivation in the PR topic) and removing this `UpdatedReplicas` check (because we no longer care if the Deployment has updated or not.  It's up to ClusterOperator to tell us if/when the component has leveled, as discussed in the PR topic).

Or are you asking specifically about the `progressingCondition == nil` case or the unrecognized `progressingCondition.Status` cases?  I think the new logic is about "if we see something we know is bad; block" and that not seeing things we think are good is no longer sufficient to block.  I'm fine logging them if we are concerned that will introduce mysterious bugs.

Reply:
My comment is a long-winded roundabout to whether the `unknown state` logging should be left as something that could be helpful since `unknown state` sounds rather scary to me. Since you're more familiar with the history of this beast I'm fine with leaving it to your judgement.

Reply:
Ok, I've pushed 9be62c45 -> cc9292a0 adding something sort of like the old `unknown state` logging.  Discussion in a new paragraph in the commit message of cc9292a0.